pageserver/src/page_cache.rs

```diff
@@ -0,0 +1,582 @@
+//
+// On locking:
+//
+// There are two levels of locking involved: There's one lock for the "mapping" from
+// page identifier (tenant ID, timeline ID, rel, block, LSN) to the buffer slot, and
+// a separate lock on each slot. To read or write the contents of a slot, you must
+// hold the lock on the slot in read or write mode, respectively. To change the mapping
+// of a slot, i.e. to evict a page or to assign a buffer for a page, you must hold
+// the mapping lock and the lock on the slot at the same time.
+//
+// Whenever you need to hold both locks simultenously, the slot lock must be acquired
+// first. This consistent ordering avoids deadlocks. To look up a page in the cache,
+// you would first look up the mapping, while holding the mapping lock, and then lock
+// the slot. You must release the mapping lock in between, to obey the lock ordering and
+// avoid deadlock.
+//
+// A slot can momentarily have invalid contents, even if it's already been inserted
+// to the mapping, but you must hold the write-lock on the slot until the contents
+// are valid. If you need to release the lock without initializing the contents,
+// you must remove the mapping first. We make that easy for the callers with
+// PageWriteGuard: when lock_for_write() returns an uninitialized page, the caller
+// must explicitly call guard.mark_valid() after it has initialized it. If the guard
+// is dropped without calling mark_valid(), the mapping is automatically removed and
+// the slot is marked free.
+//
+
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    convert::TryInto,
+    sync::{
+        atomic::{AtomicU8, AtomicUsize, Ordering},
+        RwLock, RwLockReadGuard, RwLockWriteGuard,
+    },
+};
+
+use once_cell::sync::OnceCell;
+use zenith_utils::{
+    lsn::Lsn,
+    zid::{ZTenantId, ZTimelineId},
+};
+
+use crate::{relish::RelTag, PageServerConf};
+
+static PAGE_CACHE: OnceCell<PageCache> = OnceCell::new();
+
+pub fn init(conf: &'static PageServerConf) {
+    if conf.page_cache_size == 0 {
+        return;
+    }
+
+    PAGE_CACHE.get_or_init(|| PageCache::new(conf.page_cache_size));
```

Comment:
We should copy your approach and report this as an error if called multiple times